### ISSUE TITLE
Fix tool preservation issue with XRefTargets editor

### ISSUE THREAD

**Issue by michael** [substance/scientist#35]
We have a situation where the modal (XRefTargets) closes when you select a target for the first time.

**Interestingly when you open the modal the second time everything gets preserved magically which is supposedly a bug. So before we apply the fixes described below, we should find out why components are magically preserved in this situation**. This can be reproduced on the devel branch. I've added some console.logs for various life cylcle methods so you can see what's going on.

Enable preservate rendering to fix the issue.

In substance/ui/DefaultOverlay.js add ref for each tool.

```js
  var toolName = tool.Class.static.name;
  el.append(
    $$(tool.Class, toolProps).ref(toolName)
  );
```

In XRefTool add ref for XRefTargets component

```js
    if (this.state.edit) {
      el.append(
        $$(Modal, {
          width: 'large'
        }).append(
          $$(XRefTargets, {
            node: node
          })// .ref('xrefTargets')
        )
      );
    }
```

**Comment by oliver----** [comment]
I think everything is fine with component.js.

For my investigation I activated the ref in `ui/DefaultOverlay`, but in `XRefTool` leaving `XRefTargets` without a ref. What we expect is, that anytime `XRefTool` is rerendered, `XRefTargets` get created from scratch (not preserved). 

Looking at the simple example, the following happens:

1. Putting the cursor into ref:
  ![image](https://cloud.githubusercontent.com/assets/284099/16487478/73855100-3ecb-11e6-8ddb-39e386a0cb24.png)
  ![image](https://cloud.githubusercontent.com/assets/284099/16487488/848d0e2a-3ecb-11e6-9b71-a8dafec2b04a.png)

  The tool is rendered, without state, thus no `XRefTargets`.

2. Opening the targets dialog:
  ![image](https://cloud.githubusercontent.com/assets/284099/16487875/b130b60a-3ecd-11e6-83ce-f36a83bf1612.png)

  Now `XRefTool` has `state.edit=true`, thus creating `XRefTargets` with component id `69`.

3. Adding a target:
  ![image](https://cloud.githubusercontent.com/assets/284099/16487738/f141ed32-3ecc-11e6-8062-6ae953da7cd8.png)
  ![image](https://cloud.githubusercontent.com/assets/284099/16487915/ed36c9e6-3ecd-11e6-85cb-6feb5ed530bd.png)


  Note that:
    1. `XRefTool` is re-rendered. This makes sense if we assume tools get updated on every change.
    2. A new `XRefTargets` gets created from scratch
    3. The old `XRefTargets` instance gets disposed

4. Adding another target:
  Here nothing new appears in the log, which means that `XRefTool` is not re-rendered in first place.

**Comment by oliver----** [comment]
`Component.js` is working as expected, but we need to find out why Tools don't get updated, or why it happens the first time, but not the second time.

**Comment by oliver----** [comment]
Ok. Reason is, that `ui/Overlay#shouldRerender()` is being smart. I think I added it to reduce flickering.
For now we should enable the refs, remove `ui/Overlay#shouldRerender()` and maybe optimize later.

**Comment by oliver----** [comment]
Fixed via substance/substance#712 and #36.

**Comment by michael** [comment]
Thanks for the investigation!
